### PR TITLE
Eliminate busy wait loops for game thread synchronization

### PR DESCRIPTION
**Pull request by freakode**
Uses an EventWaitHandle to block threads waiting for other threads to finish initializing.  On game shutdown uses Thread.Join() to wait for threads in the process of terminating.

### CODE REVIEW

File: osu.Framework/Platform/BasicGameHost.cs [hunk 1]
```diff
@@ -404,8 +401,8 @@ protected virtual void LoadGame(BaseGame game)
             Task.Run(delegate
             {
                 // Make sure we are not loading anything game-related before our threads have been initialized.
-                while (!ReadyToLoad)
```
Comment (peppy) [inline osu.Framework/Platform/BasicGameHost.cs]: Underlying property is still present and likely was only used here.

File: osu.Framework/Threading/GameThread.cs [hunk 2]
```diff
@@ -88,13 +90,19 @@ public GameThread(Action onNewFrame, string threadName)
             Scheduler = new Scheduler(null);
         }
 
+        public void WaitTillInitialized()
```
Comment (peppy) [inline osu.Framework/Threading/GameThread.cs]: Can we use "Until" instead of "Till" please?

File: osu.Framework/Platform/BasicGameHost.cs (hunk 1 above)
Comment (freakode) [inline osu.Framework/Platform/BasicGameHost.cs]: That is true.  So I should go ahead and nuke it?

File: osu.Framework/Threading/GameThread.cs [hunk 3]
```diff
@@ -88,13 +90,19 @@ public GameThread(Action onNewFrame, string threadName)
...
+            initializedEvent.WaitOne();
+        }
+
         private void runWork()
         {
             Scheduler.SetCurrentThread();
 
             OnThreadStart?.Invoke();
 
             isInitialized = true;
```
Comment (freakode) [inline osu.Framework/Threading/GameThread.cs]: Right now there is no remaining code that uses this bool.  We don't need to keep it unless there is some reason we want to check if the thread is initialized without blocking.

File: osu.Framework/Threading/GameThread.cs (hunk 2 above)
Comment (freakode) [inline osu.Framework/Threading/GameThread.cs]: Sure, no problem.

**Comment by peppy** [comment]
This PR breaks osu headless mode (see unit tests in osu project).

**Comment by freakode** [comment]
Alright, I'll take a look at the test case failures.

**Comment by freakode** [comment]
The test cases were failing because HeadlessGameHost depended on an overridden version of the ReadyToLoad property which was accidentally removed.  Equivalent functionality has been re-added and the failing test cases have been verified to work.  Tests in OszArchiveReaderTest fail but that is not related to this code and I know the fix.

File: osu.Framework.Desktop/Platform/HeadlessGameHost.cs [hunk 4]
```diff
@@ -35,5 +35,9 @@ protected override void DrawFrame()
         }
 
         public override IEnumerable<InputHandler> GetInputHandlers() => new InputHandler[] { };
+
+        protected override void WaitTillReadyToLoad()
```
Comment (peppy) [inline osu.Framework.Desktop/Platform/HeadlessGameHost.cs]: until? :P

Comment (freakode) [inline osu.Framework.Desktop/Platform/HeadlessGameHost.cs]: Gah, old habits die hard i suppose.

**Pull request merged**